### PR TITLE
Replicate pr 991

### PR DESCRIPTION
**Pull request by dwertent**

### CODE REVIEW

**Comment by dwertent** [comment]
@greptile

File: core/go/internal/sequencer/coordinator/transaction/ready_for_dispatch.go [hunk 1]
```diff
@@ -21,26 +21,69 @@ import (
...
+			for _, constraint := range endorsement.Constraints {
+				if constraint == prototk.AttestationResult_ENDORSER_MUST_SUBMIT {
+					t.Signer = endorsement.Verifier.Lookup
+					t.dynamicSigningIdentity = false
+					log.L(context.Background()).Debugf("Setting transaction %s signer %s based on endorsement constraint", t.ID.String(), t.Signer)
+					return
+				}
+			}
+		}
+	}
```
Comment (dwertent) [inline core/go/internal/sequencer/coordinator/transaction/ready_for_dispatch.go:43]: [cursor-auto-review] These log statements use `context.Background()` instead of the `ctx` parameter available via the method receiver or passed to the calling function (`hasDependenciesNotReady(ctx)`). This loses correlation IDs, trace spans, and other contextual metadata. Same issue in `dependentsMustWait` at lines 43, 50, 55, 60.

```go
// Before
log.L(context.Background()).Tracef(...)

// After
log.L(ctx).Tracef(...)
```

File: core/go/internal/txmgr/dependency_listener.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+func (tm *txManager) BlockedByDependencies(ctx context.Context, tx *components.ValidatedTransaction) (bool, error) {
+	for _, dep := range tx.DependsOn {
+		depTXReceipt, err := tm.GetTransactionReceiptByID(ctx, dep)
+		if err != nil {
+			// Fail safe - if we can't check the status of a dependency we should assume we're waiting for it.
+			return true, err
+		}
+		if depTXReceipt == nil || !depTXReceipt.Success {
+			// Dependency TX receipt isn't present or it failed. Either way the caller must not progress the TX in question
```
Comment (dwertent) [inline core/go/internal/txmgr/dependency_listener.go:78]: [cursor-auto-review] `BlockedByDependencies` returns `(true, nil)` for both cases: dependency receipt missing (pending) and dependency receipt present but failed. The caller (`HandleNewTx` / `HandleTxResume`) treats `blocked == true` as "silently return nil" and relies on two resumption paths: (1) periodic poll, (2) dependency listener on receipt finalization.

But if the dependent TX is created _after_ the dependency has already been finalized as failed, neither path works:
- The periodic poll calls `BlockedByDependencies` again, gets `(true, nil)`, and does nothing.
- The dependency listener fires on `FinalizeTransactions`, which already happened before this TX existed.

The TX will remain pending forever. Fix by distinguishing the two cases:

```go
if depTXReceipt == nil {
    // Still pending — block and wait for notification
    return true, nil
}
if !depTXReceipt.Success {
    // Already failed — proactively fail the dependent TX
    return true, fmt.Errorf("dependency %s already failed", dep)
}
```

Then the caller can finalize the TX as failed instead of silently dropping it.

File: core/go/internal/sequencer/coordinator/transaction/state_machine.go [hunk 3]
```diff
@@ -277,6 +295,11 @@ func init() {
 							To: State_SubmissionPrepared,
 						}},
 				},
+				Event_DependencyReverted: {
+					Transitions: []Transition{{
+						To: State_Pooled,
+					}},
+				},
```
Comment (dwertent) [inline core/go/internal/sequencer/coordinator/transaction/state_machine.go:302]: [cursor-auto-review] The `DependencyReverted` → `State_Pooled` transition is added for `State_Dispatched`, `State_SubmissionPrepared`, and `State_Submitted`. At these states, the transaction has already been handed to the public transaction manager for base-ledger submission.

Re-pooling will re-assemble and potentially dispatch a second base-ledger transaction for the same private TX. The original dispatch is still in-flight. While the original will likely revert on-chain (its dependency's state is gone), there is a race window where:
- The original dispatch succeeds before the dependency revert is reflected on-chain
- The re-assembled dispatch also succeeds, causing double execution

If the base ledger revert is guaranteed before re-dispatch (e.g. because the privacy group enforces it), this is safe. But there is no code cancelling the in-flight public TX submission. If performance or gas-waste is a concern, consider cancelling the pending public TX before re-pooling from these states, or at minimum document why double-dispatch is safe.

File: core/go/internal/sequencer/coordinator/transaction/pooled.go [hunk 4]
```diff
@@ -112,18 +112,57 @@ func (t *Transaction) initializeDependencies(ctx context.Context) error {
 		}
 	}
 
+	// Reset anything that might have been updated during an initial attempt to assembly, endorse and dispatch this TX. This is a no-op if this is the first
+	// and only time we pool & assemble this transaction but if we're re-pooling for any reason we must clear the post-assembly and any post-assembly
+	// dependencies from a previous version of the grapher.
+	t.PostAssembly = nil
+	t.dependencies = &pldapi.TransactionDependencies{}
```
Comment (dwertent) [inline core/go/internal/sequencer/coordinator/transaction/pooled.go:119]: [cursor-auto-review] `initializeForNewAssemply` resets `t.dependencies = &pldapi.TransactionDependencies{}` and calls `t.grapher.ForgetMints(t.ID)`, but does NOT re-pool its own dependents (`PrereqOf`). This breaks transitive dependency chains.

Consider A → B → C. If A reverts, `rePoolDependents` re-pools B. B transitions to `State_Pooled`, calling `initializeForNewAssemply` which clears B's `PrereqOf` (erasing the B→C link) and forgets B's mints. B eventually gets re-assembled with new output states. But:

1. C's `DependsOn` still references B by ID, so C waits for B.
2. B's `calculatePostAssembleDependencies` links new dependents based on new output states — C's input states reference B's old outputs, so the B→C `PrereqOf` link is NOT re-established.
3. When B reaches `Ready_For_Dispatch`, `notifyDependentsOfReadinessAndQueueForDispatch` iterates `B.dependencies.PrereqOf`, which no longer includes C.
4. C is permanently stuck in `Blocked` waiting for a `DependencyReady` event that will never arrive.

Fix: before resetting `t.dependencies`, iterate `t.dependencies.PrereqOf` and send `DependencyRevertedEvent` to each (same as `rePoolDependents`). This cascades the re-pool to all transitive dependents.

**Comment by dwertent** [comment]
[cursor-auto-review] 4 findings posted as inline comments. Summary below.

## Blocking issues

**1. `hasDependenciesNotReady` will permanently block transactions after dependency cleanup** — `ready_for_dispatch.go` changed the behavior when a dependency is not found in the in-memory grapher from `continue` (assume confirmed) to `return true` (block). But the coordinator cleans up confirmed transactions via `grapher.Forget()`. Once a confirmed dependency is cleaned up, its dependents will be blocked forever. This is the most critical bug in the PR — it directly contradicts the lifecycle of the grapher. Restore the previous `continue` behavior (treating missing-in-memory as confirmed), or keep confirmed transactions in the grapher until all their dependents have passed dependency checks.

**2. `BlockedByDependencies` conflates "pending" and "failed" dependencies** — Returns `(true, nil)` for both, making it impossible for the caller to fail a TX whose dependency has already failed. If a dependent TX is created after its dependency has already been finalized as failed, no event will fire to propagate the failure, and the TX will stay pending forever. The periodic poll will keep calling `BlockedByDependencies`, keep getting `(true, nil)`, and keep doing nothing. (Inline comment with suggested fix.)

**3. Transitive dependency chain broken on re-pool** — `initializeForNewAssemply` resets `t.dependencies` (clearing `PrereqOf`) and forgets mints without re-pooling its own dependents. In an A→B→C chain, reverting A re-pools B, but C is never notified. C ends up permanently stuck because B's `PrereqOf` no longer includes C after re-pooling, and B's new output states don't match C's input state references. (Inline comment with full trace and fix.)

**4. `FinalizeTransactions` → `notifyDependentTransactions` → `FinalizeTransactions` recursion** — The `gorm.go` pre-commit loop now evaluates `len(tx.preCommits)` dynamically, so new pre-commits appended during iteration are executed in the same commit. `notifyDependentTransactions` calls `FinalizeTransactions` for failed-dependency propagation, which appends another pre-commit calling `notifyDependentTransactions`. With circular `DependsOn` edges (no validation prevents them) or re-visiting already-finalized TXs (`OnConflict: DoNothing` suppresses duplicate errors), this is an unbounded loop. Add a visited set, a depth limit, or move the failure-receipt insertion directly into the pre-commit instead of calling `FinalizeTransactions` recursively.

## Design concerns

**Re-pooling from dispatched/submitted states** — `DependencyReverted` transitions from `State_Dispatched`, `State_SubmissionPrepared`, and `State_Submitted` back to `State_Pooled`. At these points, a base-ledger transaction is in-flight. The code does not cancel the pending public TX submission. While the in-flight TX will likely revert on-chain, there is a window for double execution. At minimum this warrants a code comment explaining why it is safe. (Inline comment posted.)

**Signing identity logic moved earlier but `dependentsMustWait` uses `context.Background()`** — Multiple log statements in `ready_for_dispatch.go` and `dependentsMustWait` use `context.Background()` instead of the available `ctx`, losing all trace/correlation metadata. Similarly, `notifyDependentTransactions` uses `tm.bgCtx` instead of the provided `ctx` for logging and DB queries.

## Overall assessment

This PR is not ready to merge. The dependency handling has at least three correctness bugs that can cause transactions to be permanently stuck (items 1, 2, 3) and one reliability issue that can cause an infinite loop during commit (item 4). The core feature — transaction dependency tracking, re-pooling on revert, and signing identity dispatch control — is well-architected and the test coverage is extensive (2600+ lines). But the edge cases around cleanup, transitive dependencies, and failure propagation need to be resolved before this can ship.
